File: solutions/ruby/custom-set/3/custom_set.rb
class CustomSet
  def initialize(elements = [])
    @members = {}

    elements.each { |e| add(e) }
  end

  def empty?
    @members.size.zero? # or simply @members.empty?
  end

  def member?(element)
    @members.include? element
  end

  def subset?(other)
    @members.all? { |member, _| other.member? member }
  end

  def disjoint?(other)
    @members.none? { |member, _| other.member? member }
  end

  def size
    @members.size
  end

  def ==(other)
    size == other.size && subset?(other)
  end

  def each(&block)
    return to_enum unless block_given?

    @members.each_key(&block)

    self
  end

  def add(element)
    @members[element] = :present
    self
  end

  def intersection(other)
    each.with_object(CustomSet.new) { |member, result| result.add(member) if other.member? member }
  end

  def difference(other)
    each.with_object(CustomSet.new) { |member, result| result.add(member) unless other.member? member }
  end

  def union(other)
    CustomSet.new(each + other.each)
  end  
end

File: solutions/javascript/custom-set/1/custom-set.js
const uniq = array => array.filter((item, i) => !array.includes(item, i + 1)); 

export class CustomSet {
  #elements;
  
  constructor(elements = []) {
    this.#elements = uniq(elements);
  }

  empty() {
    return this.#elements.length === 0;
  }

  contains(element) {
    return this.#elements.includes(element);
  }

  add(element) {
    if (!this.contains(element)) this.#elements.push(element);
    return this;
  }

  subset(superset) {
    return this.#elements.every(superset.contains.bind(superset));
  }

  disjoint(other) {
    return !this.#elements.some(other.contains.bind(other));
  }

  eql(other) {
    return this.subset(other) && other.subset(this);
  }

  union(other) {
    return new CustomSet(this.#elements.concat(other.#elements));
  }

  intersection(other) {
    return new CustomSet(this.#elements.filter(other.contains.bind(other)));
  }

  difference(other) {
    return new CustomSet(this.#elements.filter(element => !other.contains(element)));
  }
}
